parity with UI

devidask27 opened the pull request:
## ISSUE(S):
<!--- Links to JIRA tickets -->
<!--- [FT-XXXX](link.to/FT-XXXX): Subject of FT-XXXX -->
https://platform9.atlassian.net/browse/FT-337

## SUMMARY
<!--- Describe the change below -->

<!--- include "Fixes #FT-XXX" if you have a relevant Jira ticket -->
Currently, the UI provides some more feature options which are not present in the CLI bootstrap command. The CLI should have parity in terms of supported flags with UI.

## ISSUE TYPE
<!--- To checkmark and select applicable option(s), edit [ ] to [x]  -->
<!--- Alternatively, just delete options that do not apply and remove [ ] checkmarks  -->
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that may cause existing functionality to not work as expected)
- [] This change requires a documentation update

## IMPACTED FEATURES/COMPONENTS:
<!--- List of impacted features/components due to this change -->
<!--- delete section if not relevant -->
Impacted commands
bootstrap

## TESTING DONE

#### Manual
<!--- Please list down various use case which were part of manual testing.-->

**Current bootstrap cluster**
<img width="1258" alt="Screenshot 2022-02-25 at 12 38 19 PM" src="https://user-images.githubusercontent.com/76941923/155670723-6e35d35a-2f01-4847-b716-f34024c64483.png">
**bootstrap cluster after vs UI cluster**
<img width="1258" alt="Screenshot 2022-02-25 at 12 21 15 PM" src="https://user-images.githubusercontent.com/76941923/155670977-41b03636-8d06-41f3-aecd-5caf93482feb.png">

<img width="1258" alt="Screenshot 2022-02-25 at 12 34 13 PM" src="https://user-images.githubusercontent.com/76941923/155671084-82beb28c-bd1c-4a75-8040-a6edada2fead.png">
<img width="860" alt="Screenshot 2022-02-25 at 12 21 34 PM" src="https://user-images.githubusercontent.com/76941923/155671123-862a9c0d-f323-4077-8142-280e17fcc5bb.png">

**Incorrect pmk version**

<img width="869" alt="Screenshot 2022-03-04 at 11 58 32 AM" src="https://user-images.githubusercontent.com/76941923/156711792-ccdf7b7c-3cec-4ec5-bdfd-cfff3e56f580.png">



#### Reviewers
<!--- Add reviewers by appending their github usernames after "/cc" -->
<!--- delete section if not relevant -->
<!--- /cc @REVIEWER1_GITHUB_USERNAME, @REVIEWER2_GITHUB_USERNAME, ... --->
/cc @anupbarve @nikhilbandi

anupbarve commented on pkg/qbert/qbert.go:
```diff
@@ -419,6 +473,54 @@ func Attach_Status(attachEndpoint string, token string, byt []byte) (*http.Respo
+	var monitoringDetails string
+	var tagDetails string
+	if IsPMKversionDefined {
+		if SplitPMKversion[0] <= "1.20.11" {
```
Put the version in constants.go
Also document the logic as to why we are having such a condition wrt the version.

anupbarve commented on pkg/qbert/qbert.go:
```diff
@@ -419,6 +473,54 @@ func Attach_Status(attachEndpoint string, token string, byt []byte) (*http.Respo
+
+func getMonitoringDetails() string {
+	if !IsMonitoringDisabled {
+		Monitoring = fmt.Sprintf(`,"monitoring":"{\"retentionTime\":\"7d\"}"`)
```
Why 7 days ? 
Please document the logic

anupbarve commented on cmd/bootstrap.go:
```diff
@@ -43,35 +43,79 @@ var bootstrapCmd = &cobra.Command{
-	privileged             bool
-	allowWorkloadsOnMaster bool
-	networkPlugin          string
+	useHostName              bool
+	networkPluginOperator    bool
+	enableKubVirt            bool
+	prometheusMonitoring     bool
+	etcdBackup               bool
+	enableProfileEngine      bool
+	networkStack             int
+	apiServerFlags           []string
+	controllerManagerFlags   []string
+	schedulerFlags           []string
+	tag                      string
+	topologyManagerPolicy    string
+	reservedCPUs             string
+	containerRuntime         string
+	mtuSize                  string
+	blockSize                string
+	pmkVersion               string
+	ipEncapsulation          string
+	interfaceDetection       string
+	advancedAPIconfiguration string
+	masterVIP                string
+	masterVIPIf              string
+	metallbIPRange           string
+	containersCIDR           string
+	servicesCIDR             string
+	externalDNSName          string
+	privileged               bool
+	allowWorkloadsOnMaster   bool
+	networkPlugin            string
+	calicoNatOutgoing        int
```
All these variables need to have corresponding comments.

anupbarve commented on cmd/bootstrap.go:
```diff
@@ -80,6 +124,33 @@ func bootstrapCmdRun(cmd *cobra.Command, args []string) {
 	detachedMode := cmd.Flags().Changed("no-prompt")
 	isRemote := cmdexec.CheckRemote(bootConfig)
 
+	isEtcdBackupDisabled := cmd.Flags().Changed("etcd-backup")
+	qbert.IsMonitoringDisabled = cmd.Flags().Changed("monitoring")
+	enabledKubVirt := cmd.Flags().Changed("enable-kubeVirt")
+	if enabledKubVirt {
+		networkPluginOperator = true
+	}
+	isIPv6enabled := cmd.Flags().Changed("network-stack")
+	if isIPv6enabled {
+		useHostName = false
+		networkPlugin = "calico"
+	}
+
+	qbert.IsPMKversionDefined = cmd.Flags().Changed("pmk-version")
+	if qbert.IsPMKversionDefined {
+		//Profile Engine support check
+		qbert.SplitPMKversion = strings.Split(pmkVersion, "-")
+		if qbert.SplitPMKversion[0] < "1.20.11" {
+			containerRuntime = "docker"
+			enableProfileEngine = false
+		}
```
Move all the keywords to constants.go
Examples include, but not limited to :)
... 
calico
etcd-backup
docker
... etc

anupbarve reviewed (changes requested):
Some generic comments before this code could be reviewed.

- Please update README.md
- Please use the bug template to raise this PR in order to give more info about this change.
- Do we need to add test cases for these changes ?
- I hope you have got all the command parameters reviewed from Anmol.

devidask27 commented:
> Some generic comments before this code could be reviewed.
> 
> * Please update README.md
> * Please use the bug template to raise this PR in order to give more info about this change.
> * Do we need to add test cases for these changes ?
> * I hope you have got all the command parameters reviewed from Anmol.

Updated README.md
Added PR template
I think no need of test cases, added two functions but it only contains fmt.Sprintf functionality
Already given the binary to anmol to try it

anupbarve commented on README.md:
```diff
@@ -590,24 +590,45 @@ Usage:
-
+      --advanced-api-configuration string   Allowed API groups and version. Option: default, all & custom
+      --allow-workloads-on-master           Taint master nodes ( to enable workloads ) (default true)
+      --api-server-flags strings            Comma separated list of supported kube-apiserver flags, e.g: --request-timeout=2m0s,--kubelet-timeout=20s
+      --block-size string                   Block size determines how many Pods can run per node vs total number of nodes per cluster (default "26")
+      --container-runtime string            The container runtime for the cluster (default "containerd")
+      --containers-cidr string              CIDR for container overlay (default "10.20.0.0/16")
+      --controller-manager-flags strings    Comma separated list of supported kube-controller-manager flags, e.g: --large-cluster-size-threshold=60,--concurrent-statefulset-syncs=10
+      --enable-kubeVirt                     Enables Kubernetes to run Virtual Machines within Pods. This feature is not recommended for production workloads
+      --enable-profile-engine               Simplfy cluster governance using the Platform9 Profile Engine (default true)
+      --etcd-backup                         Enable automated etcd backups on this cluster (default true)
+      --external-dns-name string            External DNS for master VIP
+  -h, --help                                help for bootstrap
+      --http-proxy string                   Specify the HTTP proxy for this cluster. Format-> <scheme>://<username>:<password>@<host>:<port>, username and password are optional.
+      --interface-detction-method string    Interface detection method for Calico CNI (default "first-found")
+  -i, --ip strings                          IP address of the host to be prepared
+      --ip-encapsulation string             Encapsulates POD traffic in IP-in-IP between nodes (default "Always")
+      --master-virtual-interface string     Physical interface for virtual IP association
+      --master-virtual-ip string            Virtual IP address for cluster
+      --metallb-ip-range string             Ip range for MetalLB
+      --mfa string                          MFA token
+      --monitoring                          Enable monitoring for this cluster (default true)
+      --mtu-size string                     Maximum Transmission Unit (MTU) for the interface (default "1440")
+      --nat int                             Packets destined outside the POD network will be SNAT'd using the node's IP (default 1)
+      --network-plugin string               Specify network plugin ( Possible values: flannel or calico ) (default "calico")
+      --network-plugin-operator             Will deploy Platform9 CRDs to enable multiple CNIs and features such as SR-IOV
+      --network-stack int                   0 for ipv4 and 1 for ipv6
+  -p, --password string                     Ssh password for the node (use 'single quotes' to pass password)
+      --pmk-version string                  Kubernetes pmk version
+      --privileged                          Enable privileged mode for K8s API. Default: true (default true)
+  -r, --remove-existing-pkgs                Will remove previous installation if found (default false)
+      --reserved-cpu string                 Comma separated list of CPUs to be reserved for the system, e.g: 4-8,9-12
+      --scheduler-flags strings             Comma separated list of supported Kube-scheduler flags, e.g: --kube-api-burst=120,--log_file_max_size=3000
+      --services-cidr string                CIDR for services overlay (default "10.21.0.0/16")
+  -s, --ssh-key string                      Ssh key file for connecting to the node
+  -e, --sudo-pass string                    Sudo password for user on remote host
+      --tag string                          Add tag metadata to this cluster (key=value)
+      --topology-manager-policy string      Topology manager policy (default "none")
+      --use-hostname                        Use node hostname for cluster creation
+  -u, --user string                         Ssh username for the node
```
IMHO, user is going to get overwhelmed after looking at this list of options. 
We should provide examples of how to get a cluster bootstrapped with default options.
These examples should also be part of the usage.

devidask27 commented:
@anupbarve Tried to Resolve points you mentioned

- groped flags as optional and required
- added bootstrap command example in command usage message
<img width="1283" alt="Screenshot 2022-03-14 at 12 52 48 PM" src="https://user-images.githubusercontent.com/76941923/158123747-8baf120b-4d0a-41e3-9344-cafe8b175338.png">

anupbarve commented on README.md:
```diff
@@ -587,9 +587,11 @@ Node decommissioning started....This may take a few minutes....Check the latest
 
 Usage:
-  pf9ctl bootstrap [flags] cluster-name
+  pf9ctl bootstrap [flags] cluster-name, e.g: pf9ctl bootstrap testCluster --pmk-version 1.21.3-pmk.72
```
These versions keep getting changed. How would the user know which version to use ?

devidask27 replied:
we are fetching all versions first, if user have not passed any version or he entered wrong version then we are listing all versions and asking to select one. 
we can remove this hardcoded version and can make it `pf9ctl bootstrap testCluster --pmk-version <version>`

anupbarve replied:
yes, please remove the hardcoded version string from here and below as well.

devidask27 replied:
Done
